File: answers/9/main.cpp
#include <vector>
#include <chrono>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <unistd.h>
#include <cmath>

// チャンネルごとに最大値を計算
void filter_gaussian(const cv::Mat &src, cv::Mat &dst, cv::Size ksize=cv::Size(3, 3), float sigma=1.3) {
    dst = src.clone();
    cv::Mat tmp = src.clone();
    tmp.convertTo(tmp, CV_32FC3);

    int w = src.size().width;
    int h = src.size().height;

    // prepare kernel
    cv::Mat k = cv::Mat::zeros(ksize, CV_32F);
    float tot_k = 0;
    for (int x=0; x<ksize.width; x++) {
        for (int y=0; y<ksize.height; y++) {
            k.at<float>(y, x) =
                1. / (sqrt(2. * M_PI) * pow(sigma, 2))
                * exp(
                    -(pow((x - (int)(ksize.width / 2)), 2) +
                      pow((y - (int)(ksize.height / 2)), 2))
                    / (2 * pow(sigma, 2)));
            tot_k += k.at<float>(y, x);
        }
    }
    k /= tot_k;

    dst.forEach<cv::Vec3b>([&](cv::Vec3b &pixel, const int pos[]) -> void {
        int x = pos[1];
        int y = pos[0];
        cv::Vec3f value = cv::Vec3f(0, 0, 0);
        for (int step_x=0; step_x<ksize.width; step_x++) {
            for (int step_y=0; step_y<ksize.height; step_y++) {
                int col = x + step_x - (int)(ksize.width / 2);
                int row = y + step_y - (int)(ksize.height / 2);
                if (col < 0) col = 0;
                if (col >= w) col = w - 1;
                if (row < 0) row = 0;
                if (row >= h) row = h - 1;
                for (int ch=0; ch<3; ch++) {
                    value[ch] += tmp.at<cv::Vec3f>(row, col)[ch] * k.at<float>(step_y, step_x);
                }
            }
        }
        pixel = value;
    });
    dst.convertTo(dst, CV_8UC3);
}

int main() {
    cv::Mat img_orig = cv::imread("questions/dataset/images/imori_256x256_noise.png");
    cv::Mat img_result;
    filter_gaussian(img_orig, img_result, cv::Size(5, 5), 10.3);
    cv::imshow("cpp_origin", img_orig);
    cv::imshow("cpp_result", img_result);
    cv::moveWindow("cpp_result", img_orig.size().width, 0);
    cv::waitKey(3000);
    return 0;
}

File: answers/9/main.py
import numpy as np
import cv2

def filter_gaussian(img, ksize=(3, 3), sigma=1.3):
    _img = img.copy().astype(np.float32)
    ksize_h, ksize_w = ksize

    # padding
    h, w = img.shape[:2]
    pad_top, pad_bottom = ksize_h, ksize_h
    pad_left, pad_right = ksize_w, ksize_w

    _img = np.pad(_img, [(pad_top, pad_bottom), (pad_left, pad_right), (0, 0)], 'edge')
    out = np.zeros_like(_img)

    new_h, new_w = out.shape[:2]
    c = 1 if len(out.shape) == 2 else out.shape[2]

    # prepare kernel
    k = np.zeros([ksize_h, ksize_w])
    for iy in range(ksize_h):
        for ix in range(ksize_w):
            k[iy, ix] = 1 / (np.sqrt(2 * np.pi) * (sigma ** 2)) * np.exp(- ((ix - ksize_w // 2) ** 2 + (iy - ksize_h // 2) ** 2) / (2 * sigma ** 2))
    k /= k.sum()

    # filtering
    for iy in range(new_h - ksize_h):
        for ix in range(new_w - ksize_w):
            for ic in range(c):
                out[iy, ix, ic] = np.sum(_img[iy : iy + ksize_h, ix : ix + ksize_w, ic] * k)

    out = out[pad_top : pad_top + h, pad_left : pad_left + w]
    return np.clip(out, 0, 255).astype(np.uint8)

if __name__ == '__main__':
    img_orig = cv2.imread('questions/dataset/images/imori_256x256_noise.png')
    img_result = filter_gaussian(img_orig, (5, 5), 10.3)
    cv2.imshow('python_origin', img_orig)
    cv2.imshow('python_result', img_result)
    cv2.moveWindow('python_result', img_orig.shape[1], 0)
    cv2.waitKey(3000)
